include/range/v3/view/reverse.hpp

```diff
@@ -42,152 +42,201 @@ namespace ranges
+                using reverse_view_t = meta::const_if_c<const_iterable, reverse_view>;
+                using base_range_t = meta::const_if_c<const_iterable, Rng>;
 #ifndef NDEBUG
                 reverse_view_t *rng_;
 #endif
             public:
                 adaptor() = default;
-                adaptor(reverse_view_t &rng)
 #ifndef NDEBUG
+                constexpr adaptor(reverse_view_t &rng) noexcept
                   : rng_(&rng)
+                {}
+#else
+                constexpr adaptor(reverse_view_t &) noexcept
+                {}
 #endif
-                {
-                    (void)rng;
-                }
-                static iterator_t<Rng> begin(reverse_view_t &rng)
-                {
-                    return rng.get_end_(BoundedRange<Rng>());
-                }
-                static iterator_t<Rng> end(reverse_view_t &rng)
-                {
-                    return ranges::begin(rng.base());
-                }
-                range_reference_t<Rng> read(iterator_t<Rng> it) const
-                {
-                    return *--it;
-                }
-                void next(iterator_t<Rng> &it) const
+                RANGES_CXX14_CONSTEXPR static auto begin(reverse_view_t &rng)
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    rng.get_end()
+                )
+                RANGES_CXX14_CONSTEXPR static auto end(reverse_view_t &rng)
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    ranges::begin(rng.base())
+                )
+                RANGES_CXX14_CONSTEXPR auto read(iterator_t<Rng> it) const
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    *--it
+                )
+                RANGES_CXX14_CONSTEXPR void next(iterator_t<Rng> &it) const
+                    noexcept(noexcept(
+                        --it != ranges::begin(std::declval<base_range_t &>())))
                 {
                     RANGES_ASSERT(it != ranges::begin(rng_->base()));
                     --it;
                 }
-                void prev(iterator_t<Rng> &it) const
+                RANGES_CXX14_CONSTEXPR void prev(iterator_t<Rng> &it) const
+                    noexcept(noexcept(
+                        ++it != ranges::begin(std::declval<base_range_t &>())))
                 {
                     RANGES_ASSERT(it != ranges::end(rng_->base()));
                     ++it;
                 }
                 CONCEPT_REQUIRES(RandomAccessRange<Rng>())
+                RANGES_CXX14_CONSTEXPR
                 void advance(iterator_t<Rng> &it, range_difference_type_t<Rng> n) const
+                    noexcept(noexcept(
+                        void(it - ranges::begin(std::declval<base_range_t &>())),
+                        std::declval<reverse_view_t &>().get_end(),
+                        ranges::advance(it, -n)))
```

Comment:
Your obsession with making _everything_ conditionally `noexcept` is making me queasy.

Reply:
Looking at this now, I'm not pleased with the choice to have the conditional noexcept reflect debugging code. I think e.g. this function should be `noexcept(noexcept(ranges::advance(it, -n)))`, and an exception thrown while checking contracts should irrecoverable.

Reply:
...alternately, you could tell me to rip some of these out. Your house, your rules.